[Split-Checkout] Hides "Save as default billing address" from guest-checkout

filipefurtad0 opened the pull request:
#### What? Why?

Closes #9022.

<!-- Explain why this change is needed and the solution you propose.
     Provide context for others to understand it. -->

Adds a spec and a fix (hopefully) on the cases in which the default address checkbox should appear in the UI.

#### What should we test?
<!-- List which features should be tested and how.
     This can be similar to the Steps to Reproduce in the issue.
     Also think of other parts of the app which could be affected
     by your change. -->

With the split-checkout feature enabled:

- As a guest
  - Add items to the cart
  - Visit checkout
  - On the Details step
  - The "Save as default billing address" should not be visible

- As a logged-in user
  - Add items to the cart
  - Visit checkout
  - On the Details step
  - The "Save as default billing address" should be visible

#### Release notes

<!-- Please select one for your PR and delete the other. -->

Changelog Category: User facing changes

<!-- Choose a pull request title above which explains your change to a
     a user of the Open Food Network app. -->

The title of the pull request will be included in the release notes.

jibees commented on spec/system/consumer/split_checkout_spec.rb:
```diff
@@ -162,12 +162,23 @@
         expect(page).to have_current_path("/checkout/details")
       end
 
+      it 'does not show the save as default bill address checkbox' do
```
I think we should also test as connected user if the checkbox is well shown, right?

filipefurtad0 replied:
I agree :+1: 

I did not add this test here explicitly as this other PR deals with the functionality of the [default bill address checkbox](https://github.com/openfoodfoundation/openfoodnetwork/pull/8960/files#diff-d09306155331299caecd70fa8c3f245c3e4c8d3b00f9a881015fafb3eb38cda8R217), as a connected user.

What do you think?

jibees replied:
Great idea!

jibees commented on app/views/split_checkout/_details.html.haml:
```diff
@@ -61,8 +61,8 @@
           = bill_address.label :state_id, t("split_checkout.step1.address.state_id.label")
           = bill_address.select :state_id, states_for_country(bill_address_country), { selected: @order.bill_address&.state_id }, { "data-dependant-select-target": "select" }
 
-      - if spree_current_user||true
```
My bad... 😖

jibees reviewed (approved):
Thanks! Only two small comments ;)

drummer83 commented:
Thanks for this @filipefurtad0!

I have compared the release v4.1.23 (as "before") with your [build #2](https://semaphoreci.com/openfoodfoundation/openfoodnetwork-2/branches/pull-request-9021/builds/2) of PR #9021. In both cases I was not logged in, so checking out as a guest. Here is what I found:

Left: before; Right: after
![grafik](https://user-images.githubusercontent.com/1790176/160695439-6d346e0c-cbd3-4161-b930-564b1ae71c15.png)

Indeed the checkbox "Save as default billing address" is not shown anymore.

I also wanted to check that:
- the checkbox is still there when logging in as a customer during checkout process
- the checkbox is still there when logging in as a customer before shopping
- the checkbox is still there when logging in as a super admin before shopping

When logging in while on the page "1 - Your details" the checkbox will remain hidden, 
- also after refreshing that page, 
- also after clicking on "Cart" and "Checkout" again,
- also after clicking on "Cart" and "Edit cart" and "Checkout" again,
- also after clicking on "Cart" and "Edit cart" and "Continue shopping" and adding a product and then "Checkout" again,
- also after checking out and starting a new order
- also after logging out and logging back in
- also after logging in as a different user

So I basically found **no way to get the checkbox back**! That's what I thought. 
But after trying in a private browser window and after deleting all cookies it seems that the checkbox is just always hidden from the beginning. I have never seen it.

So, my conclusion is that we don't see the checkbox for guest users anymore - which is great! - but we should have another look at how to display it in the cases we want it.

Side note for me to remember: Also I was not asked if I wanted to check out as a guest after logging out and starting a new order as a guest.

Moving back to In Dev.

filipefurtad0 commented:
Thank you for the detailed testing @drummer83 .

I've rebased - to get the tests on #8960 checking whether the checkbox appears as a logged in user - and removed the bit which was making it not appearing (at all!). Should be good now. Moving to code review.

Thanks!

drummer83 commented:
Ok, this time checking as a logged in user first to verify that the checkbox is really there. It is! :+1: 
- the checkbox is there when logging in as a customer during checkout process
- the checkbox is there when logging in as a customer before shopping
- the checkbox is there when logging in as a super admin before shopping

![grafik](https://user-images.githubusercontent.com/1790176/161022444-8279f48f-2b07-4c89-b1b2-4a594448660d.png)

After that checking as a guest user. The checkbox is not displayed! :+1: 
![grafik](https://user-images.githubusercontent.com/1790176/161024664-ab0b28b0-34e0-44b2-adbc-80d826307700.png)

To see the state before this PR, see my previous test above!

Additional test: Logging in while on the page '1 - Your details' will bring back the checkbox - awesome!

I think this one is now ready to go!
Thanks @filipefurtad0! :partying_face:
